client.go

```diff
@@ -0,0 +1,312 @@
+	}
+
+	req = req.WithContext(ctx)
+	resp, err := c.httpClient.Do(req)
+	dur := time.Since(ts)
+	if err != nil {
+		return dur, "", err
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	dur = time.Since(ts)
+	if err != nil {
+		return dur, "", err
+	}
+
+	// we shouldn't see this
+	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
+		var err = errors.New(string(body))
+		return 0, "", err
+	}
+
+	version := resp.Header.Get("X-Influxdb-Version")
+	return dur, version, nil
+}
+
+// Write writes metrics to a bucket, and org.  It retries intelligently.
+// If the write is too big, it retries again, after breaing the payloads into two requests.
+func (c *Client) Write(ctx context.Context, bucket, org string, m ...Metric) (err error) {
+	if c.authorization == "" {
+		return errors.New("a token is requred for a write")
+	}
+	tries := uint64(0)
+	return c.write(ctx, bucket, org, &tries, m...)
+}
+
+func (c *Client) write(ctx context.Context, bucket, org string, triesPtr *uint64, m ...Metric) error {
+	//r, w := io.Pipe()
+	buf := &bytes.Buffer{}
+	e := lp.NewEncoder(buf)
+doRequest:
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+	for i := range m {
+		if _, err := e.Encode(m[i]); err != nil {
+			return err
+		}
+	}
+	req, err := c.makeWriteRequest(bucket, org, buf)
+	if err != nil {
+		return err
+	}
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	throwAway := make([]byte, 128)
+	for err == nil {
+		_, err = resp.Body.Read(throwAway)
+	}
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusNoContent:
+	case http.StatusBadRequest:
+		resp.Body.Close()
+		return &genericRespError{
+			Code:    resp.Status,
+			Message: "line protocol poorly formed and no points were written.  Response can be used to determine the first malformed line in the body line-protocol. All data in body was rejected and not written",
+		}
+	case http.StatusUnauthorized:
+		resp.Body.Close()
+		return &genericRespError{
+			Code:    resp.Status,
+			Message: "token does not have sufficient permissions to write to this organization and bucket or the organization and bucket do not exist",
+		}
+	case http.StatusForbidden:
+		resp.Body.Close()
+		return &genericRespError{
+			Code:    resp.Status,
+			Message: "no token was sent and they are required",
+		}
+	case http.StatusTooManyRequests:
+		resp.Body.Close()
+		err = &genericRespError{
+			Code:    resp.Status,
+			Message: "token is temporarily over quota, failed more than max retries",
+		}
+		if err2 := c.backoff(triesPtr, resp, err); err2 != nil {
+			return err2
+		}
+		goto doRequest
+	case http.StatusServiceUnavailable:
+		resp.Body.Close()
+		err = &genericRespError{
+			Code:    resp.Status,
+			Message: "service is temporarily unavaliable",
+		}
+		if err2 := c.backoff(triesPtr, resp, err); err2 != nil {
+			return err2
+		}
+		goto doRequest
+	// split up entities that are just too big
+	case http.StatusRequestEntityTooLarge:
```

Comment:
Return error and let the user send less.  Hiding this could make the client needlessly inefficient, its complicated, and its likely it won't even work.  The last time I saw this it was a single field key causing the problem.